### PR TITLE
Assertion methods auto correction for `use-t-well` rule

### LINKED ISSUE #187: Skipping assertions is changing in AVA 1.0
[Issuehunt badges](https://issuehunt.io/r/avajs/eslint-plugin-ava/issues/187)

https://github.com/avajs/ava/pull/1696 changes assertion skipping from `t.skip.true()` to `t.true.skip()`.

From searching the rule documentation this impacts `no-skip-assert` and `use-t-well`.

Interestingly enough it looks as if `use-t-well` has some support for `t.true.skip()` even though it never worked.

Note that `t.skip()` no longer exists.

---

> *Note: This issue has a bounty, so it's expected that you are an experienced programmer and that you give it your best effort if you intend to tackle this. Don't forget, if applicable, to add tests, docs (double-check for typos). And don't be sloppy. Review your own diff multiple times and try to find ways to improve and simplify your code. Instead of asking too many questions, present solutions. The point of an issue bounty is to reduce my workload, not give me more. Include a 🦄 in your PR description to indicate that you've read this. Thanks for helping out 🙌* - @sindresorhus

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $60.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($60.00)
### Submitted pull Requests
- [#277 Assertion methods auto correction for `use-t-well` rule](https://issuehunt.io/r/avajs/eslint-plugin-ava/pull/277)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/avajs/eslint-plugin-ava/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### LINKED ISSUE #244: Make `use-t-well` more auto-fixable
[Issuehunt badges](https://issuehunt.io/r/avajs/eslint-plugin-ava/issues/244)

When https://github.com/avajs/eslint-plugin-ava/pull/229 is merged, `use-t-well` will be able to autofix `falsey` to `falsy` (a common misspell).

This can be expanded further by creating a small list of common misspells that can be autofixed (`t.contxt`, `t.truthey`, `t.deepequal` ...)
I don't know what's the limit to the size of that list. It could even be allowed to customize it with an option.

---

> *Note: This issue has a bounty, so it's expected that you are an experienced programmer and that you give it your best effort if you intend to tackle this. Don't forget, if applicable, to add tests, docs (double-check for typos). And don't be sloppy. Review your own diff multiple times and try to find ways to improve and simplify your code. Instead of asking too many questions, present solutions. The point of an issue bounty is to reduce my workload, not give me more. Include a 🦄 in your PR description to indicate that you've read this. Thanks for helping out 🙌* - @sindresorhus

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $60.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($60.00)
### Submitted pull Requests
- [#277 Assertion methods auto correction for use-t-well rule, makes most misspells fixable](https://issuehunt.io/r/avajs/eslint-plugin-ava/pull/277)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/avajs/eslint-plugin-ava/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by stroncium**
Fixes #244 

Simple breadth-first early terminating Levenshtein distance auto correcter for assertion methods.
Handles all samples from the issue and some more. (Some of earlier test cases have more suiting error messages and became fixable too.)

Fixes #187 

Detects and fixes multiple `skip` and `skip` being not the last in chain.

NOTE: integration tests fail for some reason, but they fail even without this patch.
:unicorn:

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#244: Make `use-t-well` more auto-fixable](https://issuehunt.io/repos/49722492/issues/244)
- [#187: Skipping assertions is changing in AVA 1.0](https://issuehunt.io/repos/49722492/issues/187)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

File: rules/use-t-well.js [hunk 1]
```diff
@@ -3,35 +3,91 @@ const {visitIf} = require('enhance-visitors');
...
+	if (node.object.type === 'MemberExpression') {
+		return getMemberNodes(node.object).concat(node.property);
+	}
 
-		return res;
-	}, initial);
+	return [node.property];
 };
 
+const correctIdentifier = name => correcter.correct(name, Math.max(0, Math.min(name.length - 2, 2)));
```
Comment (sindresorhus) [inline rules/use-t-well.js]: Not immediately clear what the `2` signifies. Can you move it into a constant or comment?

File: rules/use-t-well.js [hunk 2]
```diff
@@ -3,35 +3,91 @@ const {visitIf} = require('enhance-visitors');
 const util = require('../util');
 const createAvaRule = require('../create-ava-rule');
 
-const isMethod = name => util.executionMethods.has(name);
+class MicroCorrecter {
```
Comment (sindresorhus) [inline rules/use-t-well.js]: Code comment about what this does.

Comment (sindresorhus) [reply rules/use-t-well.js]: Any chance you could make this a reusable npm package so we don't have to maintain the logic here? Totally optional though.

**Comment by stroncium** [comment]
@sindresorhus Moved the correction code itself to `micro-spelling-correcter`(https://github.com/stroncium/micro-spelling-correcter) which packs it all together and provides simplest interface possible. No magic constants anymore.


NOTE: codecov fails for https://github.com/stroncium/micro-spelling-correcter for some reason atm, but coverage is 100%, will look into why it fails later.

**Comment by sindresorhus** [comment]
> Fixes #187

I'd like to see a test for:

> Note that t.skip() no longer exists.

**Comment by sindresorhus** [comment]
Looks good :)

**Pull request merged**